### PR TITLE
fix(datasource/metadata): delete homepage url case

### LINKED ISSUE #14820: Missing source directory for bitnami helm charts
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Some time a go i've implemented a better helm source url handling which works as expected.
But for some reason renovate ignores the `sourceDirectory` property for the links in the PR update table.

- https://github.com/renovate-reproductions/helm-chart-source-url/pull/1

### Relevant debug logs

Nothing relevant found there
<details><summary>Logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by hasanwhitesource**
## Changes

- `homepage` property should only be deleted if `sourceUrl` has been set to it.
- Delete `homepage` if its same as `sourceUrl` after "massaging" it and its not a link to a page.
- Added regex to clean preceding slashes for url joining in `pr/body`.
- Fixed a wrong test which was passing when `sourceUrl` was having double slashes, and modified a test to have a `homepage` property.

## Context

- Closes #14820 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/metadata.ts [hunk 1]
```diff
@@ -144,8 +144,11 @@ export function addMetaData(
     } else {
       delete dep.sourceUrl;
     }
-  }
 
+    if (dep.sourceUrl === dep.homepage && !dep.homepage?.endsWith('/')) {
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:148]: what is the end slash test for?

File: lib/modules/datasource/dart/__snapshots__/index.spec.ts.snap [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 exports[`modules/datasource/dart/index getReleases processes real data 1`] = `
 Object {
+  "homepage": "https://github.com/flutter/plugins/tree/master/packages/shared_preferences/shared_preferences",
```
Comment (viceice) [inline lib/modules/datasource/dart/__snapshots__/index.spec.ts.snap:5]: 🤔 shouldn't this `sourceUrl` instead of `homepage` ? ad it's a github url?

**Comment by hasanwhitesource** [comment]
As we discussed in the issue: 
- Delete homepage property if `sourceUrl` has been set to it.
- Delete homepage if its same as sourceUrl after "massaging" it and its not a link to a page.

Also fixed the tests that were mentioned and some that needed changing.

File: lib/modules/datasource/metadata.spec.ts [hunk 3]
```diff
@@ -256,4 +256,93 @@ describe('modules/datasource/metadata', () => {
...
+        {
+          version: '1.0.2',
+          releaseTimestamp: '2000-01-02T12:34:56.000Z',
+        },
+        {
+          version: '1.0.3',
+          releaseTimestamp: '2000-01-03T12:34:56.000Z',
+        },
+      ],
+      sourceUrl: 'http://github.com/foo',
```
Comment (rarkins) [inline lib/modules/datasource/metadata.spec.ts:285]: Also not valid source url

File: lib/modules/datasource/dart/__snapshots__/index.spec.ts.snap [hunk 4]
```diff
@@ -181,7 +182,7 @@ Object {
       "version": "0.5.8",
     },
   ],
-  "sourceUrl": "http://example.com",
+  "sourceUrl": "https://github.com/flutter",
```
[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/modules/datasource/dart/__snapshots__/index.spec.ts.snap:185]: Test hasn't changed result of the `sourceUrl` only thing that changed is the extra homepage property which should be here after the fix.
Changed this `sourceUrl` value cause of this comment (maybe better change it to a better name) https://github.com/renovatebot/renovate/pull/16166#discussion_r921014934

File: lib/modules/datasource/metadata.spec.ts (hunk 3 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.spec.ts:285]: This test is essentially a different case of of this test : https://github.com/renovatebot/renovate/blob/46e856418be8242ef8de6afc517f168d89310c02/lib/modules/datasource/metadata.spec.ts#L188-L200

Where `homepage` value needs to be moved to `sourceUrl`

**Comment by viceice** [comment]
> @viceice this is meant to be closing an issue you created. Is it what you had in mind?

Maybe it partially solves it, didn't reviewd yet.

File: lib/modules/datasource/metadata.spec.ts (hunk 3 above)
Comment (viceice) [reply lib/modules/datasource/metadata.spec.ts:285]:
```suggestion
      sourceUrl: 'http://github.com/foo/bar',
```

Should at least have a valid org and repo

File: lib/modules/datasource/metadata.spec.ts [hunk 5]
```diff
@@ -256,4 +256,93 @@ describe('modules/datasource/metadata', () => {
...
+        },
+      ],
+      sourceUrl: 'http://somesource.com',
+    });
+  });
+
+  it('Should delete gitlab homepage if its same as sourceUrl after massage', () => {
+    const dep = {
+      homepage: 'https://gitlab.com/meno/',
+      sourceUrl: 'https://gitlab.com/meno/',
```
Comment (viceice) [inline lib/modules/datasource/metadata.spec.ts:322]:
```suggestion
      homepage: 'https://gitlab.com/meno/repo',
      sourceUrl: 'https://gitlab.com/meno/repo',
```

should have a repo to be valid

File: lib/modules/datasource/metadata.ts [hunk 6]
```diff
@@ -131,8 +137,11 @@ export function addMetaData(
     if (parsedHomePage?.hostname.includes('github')) {
       if (!dep.sourceUrl) {
         dep.sourceUrl = dep.homepage;
+        if (urlPathDepth(dep.homepage) < 2) {
+          // remove homepage if its not a link to a path in a github repo.
+          delete dep.homepage;
+        }
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:143]: why the condition? `homepage` will be removed anyways when equal to `sourceUrl`
```suggestion
        delete dep.homepage;
```

we should always remove it

File: lib/util/url.ts [hunk 7]
```diff
@@ -106,3 +109,43 @@ export function parseLinkHeader(
   }
   return _parseLinkHeader(linkHeader);
 }
+
+// this method returns path depth of a url
+// example : https://github.com/org/repo/nested-path
+// output : 3.  (org->repo->nested-path depth of this path is 3)
+export function urlPathDepth(url: string | undefined): number {
```
Comment (viceice) [inline lib/util/url.ts:116]: I don't think we need this

File: lib/util/url.ts [hunk 8]
```diff
@@ -106,3 +109,43 @@ export function parseLinkHeader(
...
+  }
+  return githubUrlRe.test(url);
+}
+
+export function isGitLabUrl(url: string | undefined): boolean {
+  if (url === undefined) {
+    return false;
+  }
+  return gitlabUrlRe.test(url);
+}
```
Comment (viceice) [inline lib/util/url.ts:141]: Consider `detectPlatform`

https://github.com/renovatebot/renovate/blob/a12b47364f4de65087adf11e9be30b2d7b7f2565/lib/modules/platform/util.ts#L8-L37

File: lib/modules/datasource/metadata.ts [hunk 9]
```diff
@@ -170,3 +182,17 @@ export function addMetaData(
...
+  if (dep.homepage === undefined) {
+    return false;
+  }
+  if (isGitHubUrl(dep.homepage)) {
+    return dep.sourceUrl === massageGithubUrl(dep.homepage);
+  }
+  if (isGitLabUrl(dep.homepage)) {
+    return dep.sourceUrl === massageGitlabUrl(dep.homepage);
+  }
+  return dep.sourceUrl === dep.homepage;
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:197]: thinking again about this. We should not remove homepage when it's not a valid gitlab or github url.
We need `sourceUrl` for changelogs. so it doesn't matter when they are equal but don't point to a supported changelog source

check also other comment about `detectPlatform`

File: lib/modules/datasource/metadata.ts [hunk 10]
```diff
@@ -131,8 +137,11 @@ export function addMetaData(
     if (parsedHomePage?.hostname.includes('github')) {
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:137]: i think this needs another refactor to use `detectPlatform`, see other comments

File: lib/modules/datasource/metadata.ts (hunk 6 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.ts:143]: Hey @viceice if you consider the exact case of the bug that you've opened. we shouldn't always remove the `homepage` url. 
For example: 
`homepage: 'https://github.com/org/repo/path/to/something',
sourceUrl: undefined` 
We set `sourceurl` to `homepage` and after massaging `sourceurl` will be `https://github.com/org/repo` so therefore we need to keep `homepage` url set to `https://github.com/org/repo/path/to/something`.
Which is why I've created the `urlPathDepth` function
Happy to provide additional information if needed cause the fix isn't that trivial and will add more tests.

Comment (viceice) [reply lib/modules/datasource/metadata.ts:143]: then we should remove this here and compare after massaging. if they are equal, remove Homepage

File: lib/util/url.ts (hunk 7 above)
Comment (hasanwhitesource) [reply lib/util/url.ts:113]: I have made examples on why this check is needed: 
https://github.com/renovatebot/renovate/blob/2bf2f3f4b763573c0bdcb5e06e90beb93f9981af/lib/modules/datasource/metadata.ts#L206-L214

So for these cases we do not need to keep `homepage` url.

File: lib/modules/datasource/metadata.ts (hunk 9 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.ts]: I've made a check just for `gitlab` and `github` urls.

File: lib/modules/datasource/metadata.ts [hunk 11]
```diff
@@ -120,20 +121,15 @@ export function addMetaData(
   }
 
   if (
-    dep.changelogUrl?.includes('github.com') && // lgtm [js/incomplete-url-substring-sanitization]
+    dep.changelogUrl &&
+    detectPlatform(dep.changelogUrl) === 'github' && // lgtm [js/incomplete-url-substring-sanitization]
```
Comment (nabeelsaabna) [inline lib/modules/datasource/metadata.ts:125]: clean `lgtm` comment

File: lib/modules/datasource/metadata.ts [hunk 12]
```diff
@@ -120,20 +121,15 @@ export function addMetaData(
...
-  if (dep.homepage) {
-    const parsedHomePage = parseUrl(dep.homepage);
-    if (parsedHomePage?.hostname.includes('github')) {
-      if (!dep.sourceUrl) {
-        dep.sourceUrl = dep.homepage;
-      }
-      delete dep.homepage;
-    }
+  if (dep.homepage && !dep.sourceUrl) {
+    setHomepageToSourceURl(dep);
```
Comment (nabeelsaabna) [inline lib/modules/datasource/metadata.ts:132]: rename to match logic

File: lib/modules/datasource/metadata.ts [hunk 13]
```diff
@@ -170,3 +169,57 @@ export function addMetaData(
...
+export function setHomepageToSourceURl(dep: ReleaseResult): void {
+  if (dep.homepage === undefined) {
+    return;
+  }
+  const platform = detectPlatform(dep.homepage);
+  if (platform !== 'github' && platform !== 'gitlab') {
+    return;
+  }
+  dep.sourceUrl = dep.homepage;
+  if (shouldDeleteHomepage(massageUrl(dep.sourceUrl), dep.homepage)) {
```
Comment (nabeelsaabna) [inline lib/modules/datasource/metadata.ts:182]: move the `massageUrl()` logic into the `shouldDeleteHomepage` function

File: lib/modules/datasource/metadata.spec.ts [hunk 14]
```diff
@@ -256,4 +261,180 @@ describe('modules/datasource/metadata', () => {
...
+        },
+      ],
+      sourceUrl: 'https://github.com/foo/bar',
+    });
+  });
+
+  it('Should delete homepage if its same as sourceUrl after massage', () => {
+    const dep = {
+      homepage: 'http://somesource.com',
+      sourceUrl: 'http://somesource.com',
```
Comment (viceice) [inline lib/modules/datasource/metadata.spec.ts:297]: I'm not sure, but i think this causes duplicate links in PR with same taget, so one probably `homepage` should removed?

@rarkins WDYT?

File: lib/modules/datasource/metadata.ts [hunk 15]
```diff
@@ -170,3 +169,51 @@ export function addMetaData(
...
+  //   gitub.com/org/repo
+  //   github.com/org/repo/
+  if (urlPathDepth(homepage) <= 2) {
+    return true;
+  }
+  let hPath = homepageParsed.pathname;
+  if (hPath.charAt(hPath.length - 1) === '/') {
+    hPath = hPath.substring(0, hPath.length - 1); // remove last slash
+  }
+  return hPath === sourceUrlParsed.pathname;
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:218]: i think still think we can remove `homepage` when both urls (with trim trailing slash) are the same. The changelog code will do a platform check again, so this whole check doesn't matter at all and should be simplified.

Renovate will also fallback to `sourceUrl` for the PR homepage link if `homepage` is missing.

File: lib/modules/datasource/metadata.spec.ts (hunk 14 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.spec.ts:297]: Yes this tests that (duplicate links).
If both are same then only `sourceUrl` will be kept right?

File: lib/modules/datasource/metadata.ts (hunk 15 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.ts]: Removed the check and will take note of this for the future.

File: lib/modules/datasource/metadata.spec.ts (hunk 14 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.spec.ts:297]: The test name is `Should delete homepage if its same as sourceUrl after massage` 
Can you please explain further where is the problem in the test ?

Comment (viceice) [reply lib/modules/datasource/metadata.spec.ts:297]: test looks ok now

File: lib/modules/datasource/metadata.ts [hunk 16]
```diff
@@ -120,20 +121,15 @@ export function addMetaData(
...
-    const parsedHomePage = parseUrl(dep.homepage);
-    if (parsedHomePage?.hostname.includes('github')) {
-      if (!dep.sourceUrl) {
-        dep.sourceUrl = dep.homepage;
-      }
-      delete dep.homepage;
-    }
+  if (dep.homepage && !dep.sourceUrl) {
+    setSourceUrlToHomepage(dep);
   }
```
Comment (rarkins) [inline lib/modules/datasource/metadata.ts:133]: I don't like this logic and naming. "if there's a home page but no source URL then set the source URl to the home page" is not logically correct.

**Review by rarkins — CHANGES_REQUESTED** [review]
This logic flow and functions are hard to follow and will be hard to maintain

File: lib/modules/datasource/metadata.ts [hunk 17]
```diff
@@ -170,3 +169,41 @@ export function addMetaData(
...
+  }
+  const homepageParsed = parseUrl(homepage);
+  if (is.nullOrUndefined(homepageParsed)) {
+    return false;
+  }
+  let hPath = homepageParsed.pathname;
+  if (hPath.charAt(hPath.length - 1) === '/') {
+    hPath = hPath.substring(0, hPath.length - 1); // remove last slash
+  }
+  return hPath === sourceUrlParsed.pathname;
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:208]:
```suggestion
  return trimTrailingSlash(homepage) === trimTrailingSlash(sourceUrl);
```

Don't parse urls

File: lib/modules/datasource/metadata.ts (hunk 16 above)
Comment (viceice) [reply lib/modules/datasource/metadata.ts:133]: yes, bad naming 🤔

Comment (nabeelsaabna) [reply lib/modules/datasource/metadata.ts:133]: the `shouldDeleteHomepage()` can be removed from inside `setSourceUrlToHomepage()`

Comment (nabeelsaabna) [reply lib/modules/datasource/metadata.ts:133]: and keep a single place to check/delete the homepage

File: lib/modules/datasource/metadata.ts (hunk 17 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.ts]: If we straightly check without parsing then then tests having `sourceUrl: 'github.com/org/repo'` and `homepage:'www.github.com/org/repo'` will fail. That's why we need to check the path.

But I have removed the manual slash trimming now and used `trimTrailingSlash`

File: lib/modules/datasource/metadata.ts [hunk 18]
```diff
@@ -120,19 +121,17 @@ export function addMetaData(
   }
 
   if (
-    dep.changelogUrl?.includes('github.com') && // lgtm [js/incomplete-url-substring-sanitization]
+    dep.changelogUrl &&
+    detectPlatform(dep.changelogUrl) === 'github' &&
     !dep.sourceUrl
```
Comment (nabeelsaabna) [inline lib/modules/datasource/metadata.ts:126]:
```suggestion
    !dep.sourceUrl && dep.changelogUrl &&
    detectPlatform(dep.changelogUrl) === 'github'
```

Comment (nabeelsaabna) [reply lib/modules/datasource/metadata.ts:126]: change order to emphasize  `dep.sourceUrl` updated only in case it has no value

File: lib/modules/datasource/metadata.ts [hunk 19]
```diff
@@ -153,7 +152,10 @@ export function addMetaData(
         }) || dep.sourceUrl;
     }
   }
-
+  if (shouldDeleteHomepage(dep.sourceUrl, dep.homepage)) {
+    // remove homepage if its not a link to a path
```
Comment (nabeelsaabna) [inline lib/modules/datasource/metadata.ts:156]:
```suggestion
```

Comment (nabeelsaabna) [reply lib/modules/datasource/metadata.ts:156]: please add description to the `shouldDeleteHomepage` function instead

File: lib/modules/datasource/metadata.ts [hunk 20]
```diff
@@ -170,3 +171,29 @@ export function addMetaData(
     }
   }
 }
+
+// remove homepage if its not a link to a path or if its same as sourceurl.
```
Comment (nabeelsaabna) [inline lib/modules/datasource/metadata.ts:175]:
```suggestion
// remove homepage if both sourceUrl and homepage exists and:
// 1. homepage its not a link to a repository subpath
// 2. or both are the same
```

**Review by nabeelsaabna — DISMISSED** [review]

File: lib/modules/datasource/metadata.spec.ts [hunk 21]
```diff
@@ -256,4 +260,180 @@ describe('modules/datasource/metadata', () => {
...
+        },
+        {
+          version: '1.0.3',
+          releaseTimestamp: '2000-01-03T12:34:56.000Z',
+        },
+      ],
+    });
+  });
+
+  it('should delete homepage', () => {
```
Comment (viceice) [inline lib/modules/datasource/metadata.spec.ts:410]: use test.each and table syntax, so we've separate tests and better readability 

search for samples via `test.each` in repo 😉

File: lib/modules/datasource/metadata.ts [hunk 22]
```diff
@@ -170,3 +171,31 @@ export function addMetaData(
     }
   }
 }
+
+// remove homepage if both sourceUrl and homepage exists and:
+// 1. homepage its not a link to a repository subpath
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:176]: i think this isn't true, please update the docs

File: lib/modules/datasource/metadata.spec.ts [hunk 23]
```diff
@@ -256,4 +260,180 @@ describe('modules/datasource/metadata', () => {
...
+        {
+          version: '1.0.3',
+          releaseTimestamp: '2000-01-03T12:34:56.000Z',
+        },
+      ],
+      sourceUrl: 'http://somesource.com',
+    });
+  });
+
+  it('Should delete gitlab homepage if its same as sourceUrl after massage', () => {
```
[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.spec.ts]: Updated the test

File: lib/modules/datasource/metadata.ts (hunk 22 above)
Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.ts:176]: Is this more clear ? 
```suggestion
// 1. homepage is not a link to a path in repo (e.g https://github.com/org/repo/)
```
I think if the function's code is clear enough we can remove the comment above it entirely.

File: lib/modules/datasource/metadata.spec.ts [hunk 24]
```diff
@@ -256,4 +260,166 @@ describe('modules/datasource/metadata', () => {
       'https://example.com/foo/bar'
     );
   });
+
+  it('Should remove homepage and set to sourceUrl when not a link to a path in repo', () => {
```
Comment (viceice) [inline lib/modules/datasource/metadata.spec.ts:264]:
```suggestion
  it('Should remove homepage when homepage and sourceUrl are same', () => {
```

wrong test title

File: lib/modules/datasource/metadata.spec.ts [hunk 25]
```diff
@@ -256,4 +260,166 @@ describe('modules/datasource/metadata', () => {
...
+        },
+      ],
+      sourceUrl: 'https://github.com/foo/bar',
+    });
+  });
+
+  it('Should delete homepage if its same as sourceUrl after massage', () => {
+    const dep = {
+      homepage: 'http://www.somesource.com',
+      sourceUrl: 'http://somesource.com',
```
Comment (viceice) [inline lib/modules/datasource/metadata.spec.ts:296]: i don't think we should tread those as same urls 🤔

Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.spec.ts:296]: @viceice This logic wasn't newly added its what you get after the massage, there are tests that already rely on this.
For example : 

https://github.com/renovatebot/renovate/blob/8586f463ea0b182fe274c2d4c50a79792c860390/lib/modules/datasource/metadata.spec.ts#L176-L186

https://github.com/renovatebot/renovate/blob/8586f463ea0b182fe274c2d4c50a79792c860390/lib/modules/datasource/metadata.spec.ts#L188-L199

Comment (viceice) [reply lib/modules/datasource/metadata.spec.ts:296]: strange, shouldn't those only work for github and gitab urls? we can fix that in seperate pr. THose shouldn't be treated equal.

Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.spec.ts:296]: @viceice So expected result of this test should be : 
```
{
  homepage: "http://www.somesource.com",
  sourceUrl: "http://somesource.com",
  releases: [
    {
      version: "1.0.1",
      releaseTimestamp: "2000-01-01T12:34:56.000Z",
    },
    {
      version: "1.0.2",
      releaseTimestamp: "2000-01-02T12:34:56.000Z",
    },
    {
      version: "1.0.3",
      releaseTimestamp: "2000-01-03T12:34:56.000Z",
    },
  ],
}
```
?

Comment (viceice) [reply lib/modules/datasource/metadata.spec.ts:296]: I think so

Comment (hasanwhitesource) [reply lib/modules/datasource/metadata.spec.ts:296]: @viceice My bad on this, I've made a check to remove homepage explicitly for Github and Gitlab. And just to check for removal if its equal to sourceUrl when its not Github or Gitlab.

File: lib/modules/datasource/metadata.spec.ts [hunk 26]
```diff
@@ -256,4 +260,171 @@ describe('modules/datasource/metadata', () => {
       'https://example.com/foo/bar'
     );
   });
+
+  it('Should remove homepage when homepage and sourceUrl are same', () => {
```
Comment (viceice) [inline lib/modules/datasource/metadata.spec.ts:264]: still wrong title, there is no sourceUrl

File: lib/modules/datasource/metadata.spec.ts [hunk 27]
```diff
@@ -256,4 +260,171 @@ describe('modules/datasource/metadata', () => {
...
+        {
+          version: '1.0.3',
+          releaseTimestamp: '2000-01-03T12:34:56.000Z',
+        },
+      ],
+      sourceUrl: 'https://github.com/foo/bar',
+    });
+  });
+
+  it('Should delete homepage if its same as sourceUrl after massage', () => {
```
Comment (viceice) [inline lib/modules/datasource/metadata.spec.ts:293]: also wrong title

File: lib/modules/datasource/metadata.ts [hunk 28]
```diff
@@ -170,3 +171,30 @@ export function addMetaData(
     }
   }
 }
+
+// Remove homepage when:
```
Comment (viceice) [inline lib/modules/datasource/metadata.ts:175]: make  tsdoc comment

**Comment by PhilipAbed** [comment]
I'm closing this PR and opening another because
Hasan is not working on this any longer
new PR is https://github.com/renovatebot/renovate/pull/17372
